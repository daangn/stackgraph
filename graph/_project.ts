import { Reducer, Stream } from "../deps/rimbu.ts"
import { Project, SourceFile } from "../deps/ts_morph.ts"
import { denoCompilerOptions } from "../utils/project.ts"

/** In-Memory ESM Project for quick testing */
export const inMemoryProject = () =>
	new Project({
		compilerOptions: denoCompilerOptions,
		useInMemoryFileSystem: true,
	})

/**
 * Add source files to in-memory project
 */
export const withSrc =
	(project: Project) =>
	<const T extends string>(src: Record<T, string>): Record<T, SourceFile> =>
		Stream.fromObject(src)
			.map(([path, text]) =>
				[path, project.createSourceFile(path, text)] as const
			)
			.reduce(Reducer.toJSObject())
