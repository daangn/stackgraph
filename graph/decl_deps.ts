import { Stream } from "../deps/rimbu.ts"
import {
	type ClassDeclaration,
	type FunctionDeclaration,
	type Node,
	ReferencedSymbol,
	ReferencedSymbolEntry,
	SyntaxKind,
	type VariableDeclaration,
} from "../deps/ts_morph.ts"
import { encodeVSCodeURI } from "./vscode_uri.ts"

export type Declaration =
	| VariableDeclaration
	| ClassDeclaration
	| FunctionDeclaration

export type DeclDeps = Map<Declaration, Declaration[]>

/**
 * Get the topmost declaration
 *
 * @privateRemarks
 *
 * FIXME(#1): handle destructuring and multiple variable declarations
 * RFC: only track top-level declarations or inside if statements?
 * FIXME: also track `function` statements, are there other types of declarations?
 */
export const getTopDecl = (
	ref: ReferencedSymbolEntry,
): Declaration | undefined => {
	const target = ref.getNode().getAncestors()

	// 	.at(-2) // HACK: last is SourceFile, second to last is the variable/class declaration

	// return (target?.isKind(SyntaxKind.VariableDeclaration) ||
	// 		target?.isKind(SyntaxKind.ClassDeclaration))
	// 	? target
	// 	: undefined

	return target.findLast((x): x is Declaration =>
		x.isKind(SyntaxKind.VariableDeclaration) ||
		x.isKind(SyntaxKind.ClassDeclaration) ||
		x.isKind(SyntaxKind.FunctionDeclaration)
	)
}

export const equals = (a: Node, b: Node) =>
	encodeVSCodeURI(a) === encodeVSCodeURI(b)

export const getActualReferences = (
	symbol: ReferencedSymbol,
): Stream<ReferencedSymbolEntry> =>
	Stream.from(symbol
		.getReferences())
		.filter((ref) => !equals(ref.getNode(), symbol.getDefinition().getNode()))
		.filter((ref) =>
			ref.getNode().getParent()?.getKindName() !== "ImportClause"
		)

export const getReferencedDecls = (
	node: Declaration,
): Stream<Declaration> =>
	Stream.from(node.findReferences())
		.flatMap(getActualReferences)
		.map((x) => getTopDecl(x))
		.filter((x): x is Declaration => x !== undefined)

export const getDeclDeps = (links: Declaration[]): DeclDeps => {
	const graph: DeclDeps = new Map()

	let current = links
	while (current.length > 0) {
		current = current.flatMap((node) => {
			const references = getReferencedDecls(node).toArray()
			graph.set(node, references)
			return references
		})
	}
	return graph
}
