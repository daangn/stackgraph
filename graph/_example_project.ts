import { Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { inMemoryProject } from "./_project.ts"

/**
 * https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements
 *
 * Example file of all top-level javascript declarations
 */
export const declExampleText = (h: string | number = "") => /*javascript*/ `
    // https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements#declaring_variables
    export var varVar${h} = 1
    export var varVarIgnore${h} = 2

    export let varLet${h} = 1
    export let varLetIgnore${h} = 2

    export const varConst${h} = 1
    export const varConstIgnore${h} = 2

    export const varArrowFn${h} = () => {}
    export const varArrowFnIgnore${h} = () => {}

    // https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements#functions_and_classes
    export function functionFn${h}() {}
    export function functionFnIgnore${h}() {}

    export function${h}* functionGen() {}
    export function${h}* functionGenIgnore() {}

    export async function${h} functionAsync() {}
    export async function${h} functionAsyncIgnore() {}

    export async function${h}* functionAsyncGen() {}
    export async function${h}* functionAsyncGenIgnore() {}

    export class Class${h} {}
    export class ClassIgnore${h} {}

    // others
    if (true) {
        const variableInIf${h} = 3
    }
`

/**
 * Arbitarily large file with many declarations for benchmarking
 */
export const getDeclExampleFile = (amount: number) => {
	const project = inMemoryProject()

	const text = Stream
		.range({ amount }).map(declExampleText).join({ sep: "\n\n" })

	return project.createSourceFile("a.ts", text)
}

export const exampleSrc = {
	"a.ts": /*javascript*/ `
        export const a = "a"

	    export const aaa = () => a
	`,
	"b.tsx": /*javascript*/ `
        import { a, aaa } from "./a.ts"
        export { a }

        export const Comp = () => <div>{a}</div>
        export const CompAAA = () => <div>{aaa()}</div>

        // unrelated to a and should be ignored
        export const Unrelated = () => <div>Unrelated</div>
    `,
	"c.tsx": /*javascript*/ `
        import { Comp } from "./b.tsx"

        export const Page = () => <div><Comp/></div>
    `,
	"d.tsx": /*javascript*/ `
        import { a } from "./b.tsx"
        import { Page } from "./c.tsx"

        export const InnerImport = () => <Page/>
        export const AliasedImport = () => <div>{a}</div>
    `,
}
