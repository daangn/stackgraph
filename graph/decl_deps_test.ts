import { inMemoryProject, withSrc } from "./_project.ts"
import { exampleSrc } from "./_example_project.ts"
import { asNameRecord, getDeclDeps } from "./decl_deps.ts"
import { assertEquals } from "../test_deps.ts"

const project = inMemoryProject()
const { "a.ts": a } = withSrc(project)(exampleSrc)

const aDecl = a.getVariableDeclarationOrThrow("a")
const declDeps = getDeclDeps([aDecl])
const result = asNameRecord(declDeps)

Deno.test("getDeclDeps() finds references to other files", () => {
	assertEquals(result, {
		a: new Set(["Comp", "AliasedImport"]),
		Page: new Set(["InnerImport"]),
		Comp: new Set(["Page"]),
		InnerImport: new Set([]),
		AliasedImport: new Set([]),
	})
})
