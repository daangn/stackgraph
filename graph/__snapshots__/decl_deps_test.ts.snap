export const snapshot = {};

snapshot[`getDeclDeps() searches (aliased) references for a 1`] = `
{
  AliasedImport: Set(0) {},
  Comp: Set(1) {
    "Page",
  },
  CompAAA: Set(0) {},
  InnerImport: Set(0) {},
  Page: Set(1) {
    "InnerImport",
  },
  a: Set(3) {
    "AliasedImport",
    "Comp",
    "aaa",
  },
  aaa: Set(1) {
    "CompAAA",
  },
}
`;

snapshot[`getDeclDeps() handles identical export names 1`] = `
{
  "AliasedImport (/d.tsx:6:22)": Set(0) {},
  "Comp (/b.tsx:5:22)": Set(1) {
    "Page (/c.tsx:8:22)",
  },
  "CompAAA (/b.tsx:6:22)": Set(0) {},
  "InnerImport (/d.tsx:5:22)": Set(0) {},
  "Page (/c.tsx:8:22)": Set(1) {
    "InnerImport (/d.tsx:5:22)",
  },
  "Unrelated (/b.tsx:9:22)": Set(0) {},
  "a (/a.ts:2:22)": Set(3) {
    "AliasedImport (/d.tsx:6:22)",
    "Comp (/b.tsx:5:22)",
    "aaa (/a.ts:4:19)",
  },
  "a (/c.tsx:6:22)": Set(0) {},
  "aaa (/a.ts:4:19)": Set(1) {
    "CompAAA (/b.tsx:6:22)",
  },
}
`;
